### PR TITLE
chore: add validation of feature dependency rule

### PR DESCRIPTION
**Pull request by kentakozuka**
part of https://github.com/bucketeer-io/bucketeer/issues/963

### CODE REVIEW

File: pkg/feature/api/validation.go [hunk 1]
```diff
@@ -1154,7 +1158,36 @@ func validateRule(variations []*featureproto.Variation, rule *featureproto.Rule,
 		}
 		return dt.Err()
 	}
-	return validateStrategy(variations, rule.Strategy, localizer)
+	// Check depencendy.
```
Comment (cre8ivejp) [inline pkg/feature/api/validation.go:1161]:
```suggestion
	// Check dependency.
```

File: pkg/feature/api/validation.go [hunk 2]
```diff
@@ -1154,7 +1158,36 @@ func validateRule(variations []*featureproto.Variation, rule *featureproto.Rule,
...
+	tarF.Rules = append(tarF.Rules, rule)
+	defer func() { tarF.Rules = tarF.Rules[:len(tarF.Rules)-1] }()
+	err := validateFeatureDependencies(fs)
```
Comment (cre8ivejp) [inline pkg/feature/api/validation.go:1164]:
```suggestion
	if err := validateFeatureDependencies(fs); err != nil {
```

File: pkg/feature/api/validation.go [hunk 3]
```diff
@@ -1154,7 +1158,36 @@ func validateRule(variations []*featureproto.Variation, rule *featureproto.Rule,
...
+	if err != nil {
+		if err == evaluation.ErrCycleExists {
```
Comment (cre8ivejp) [inline pkg/feature/api/validation.go:1166]: The `errors.Is` is preferable to `==` operator.
```suggestion
		if errors.Is(err, evaluation.ErrCycleExists) {
```

File: pkg/feature/api/validation.go [hunk 4]
```diff
@@ -1385,10 +1418,10 @@ func validateAddPrerequisite(
 	if err := validateVariationID(fs, p, localizer); err != nil {
 		return err
 	}
-	evaluator := evaluation.NewEvaluator()
 	prevPrerequisites := tarF.Prerequisites
 	tarF.Prerequisites = append(tarF.Prerequisites, p)
-	_, err := evaluator.TopologicalSort(fs)
+	defer func() { tarF.Prerequisites = prevPrerequisites }()
+	err := validateFeatureDependencies(fs)
```
Comment (cre8ivejp) [inline pkg/feature/api/validation.go:1424]:
```suggestion
	if err := validateFeatureDependencies(fs); err != nil {
```

**Review by cre8ivejp — DISMISSED** [review]
I left a few comments regarding small things. Not really necessary.

**Comment by kentakozuka** [comment]
@cre8ivejp Fixed 👍

**Review by cre8ivejp — APPROVED** [review]
Thank you!

**Pull request merged**